### PR TITLE
Akodali/upgrade support

### PR DESCRIPTION
**Pull request by akodali18**
Auto Upgrade support for proxy and collector

### CODE REVIEW

File: install/wavefront-operator/values.yaml [hunk 1]
```diff
@@ -36,6 +36,11 @@ collector:
   ## Setting this to true is strongly recommended
   useDaemonset: true
 
+  ## If set to true, Collector pods will be upgraded automatically in case new minor upgrade version is available.
+  ## For pinning Collector to a specific version, you will need to set this option to false.
+  ## We support only minor version Auto Upgrades.
+  # enableAutoUpgrade: true
```
Comment (vikramraman) [inline install/wavefront-operator/values.yaml]: While you're at it, can you please change the `collector.image.tag` to version `1.0.3`.

File: pkg/controller/wavefrontupgradeutil/upgrade.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
+package wavefrontupgradeutil
```
Comment (vikramraman) [inline pkg/controller/wavefrontupgradeutil/upgrade.go]: Can we just name this package `util`? We can add other common stuff under here over time.

File: pkg/controller/wavefrontupgradeutil/upgrade.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	"github.com/Masterminds/semver"
+)
+
+const (
+	ProxyImageName = "proxy"
+
+	CollectorImageName = "wavefront-kubernetes-collector"
+)
+
+func GetLatestVersion(crImageName string, currentVersion string, enableAutoUpgrade bool) (string, error) {
```
Comment (vikramraman) [inline pkg/controller/wavefrontupgradeutil/upgrade.go]: Can you add some unit tests around this? And also add a `.travis.yml` file for this repo that runs the unit tests.

File: pkg/controller/wavefrontcollector/wavefrontcollector_controller.go [hunk 4]
```diff
@@ -105,19 +108,46 @@ func (r *ReconcileWavefrontCollector) Reconcile(request reconcile.Request) (reco
...
+		if err != nil {
+			reqLogger.Error(err, "Failed to update WavefrontCollector CR status")
+			return reconcile.Result{}, err
+		}
+		reqLogger.Info("Updated WavefrontCollector CR Status.")
+	}
+	return reconcile.Result{RequeueAfter: 1 * time.Hour}, nil
+}
+
+func getlatestCollector(reqLogger logr.Logger, instance *wavefrontv1alpha1.WavefrontCollector) {
```
Comment (vikramraman) [inline pkg/controller/wavefrontcollector/wavefrontcollector_controller.go]: nit: camel case `getLatestCollector`

File: pkg/controller/util/upgrade.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+		return currentVersion, nil
+	}
+
+	// "latest" effectively renders auto upgrade useless.
+	if currentVersion == "latest" {
+		return currentVersion, nil
+	}
+
+	// The last 20 tags should be good. Don't expect customers to be using a really old version of CR.
+	url := "https://registry.hub.docker.com/v2/repositories/wavefronthq/" + crImageName + "/tags/?page_size=20"
```
Comment (vikramraman) [inline pkg/controller/util/upgrade.go]: Should we add some checks against the image name / location before doing this? Account for openshift / other custom locations?

File: pkg/controller/util/upgrade.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+		return "", fmt.Errorf("unexpected http GET status: %s", resp.Status)
+	}
+	contents, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+
+	// The below will get us the versions from json.
+	// Ex for Proxy: [latest 5.5 5.1 4.38 4.36 4.35 4.34 4.33 4.32 4.31]
+	versions := gjson.Get(string(contents), "results.#.name")
```
Comment (vikramraman) [inline pkg/controller/util/upgrade.go]: Break this into a separate function that processes the result of the call to the docker hub API and add unit tests around this by passing in contents with different image tags etc. For example if we tag as `v1.0.3` vs `1.0.3` and such.

Comment (akodali18) [reply pkg/controller/util/upgrade.go]: This code just unmarshall's the docker "tag" string. I don't think unit testing this is needed(gjson.get() does nothing different based on the value of the tag itself)

As a side note, The Masterminds/semver library we use to compare tags support only comparing semantically versioned tags. If we tag our docker images with non semantic versioned tags (Ex: version1.0.3) , our current auto upgrade comparison will not work. 
An exception is when Tag is prefixed "v" . The library does work optionally with any semantic versioned tag prefix with "v" - so currently comparing "v1.0.3" and "1.0.3" will both work.

**Pull request merged**